src/meta-srv/src/handler.rs

```diff
@@ -103,13 +114,229 @@ impl HeartbeatHandlerGroup {
         let mut acc = HeartbeatAccumulator::default();
         let handlers = self.handlers.read().await;
         for h in handlers.iter() {
-            h.handle(&req, &mut ctx, &mut acc).await?;
+            if ctx.is_skip_all() {
+                break;
+            }
+
+            let role = req
+                .header
+                .as_ref()
+                .and_then(|h| Role::from_i32(h.role))
+                .context(error::InvalidArgumentsSnafu {
+                    err_msg: format!("invalid role: {:?}", req.header),
+                })?;
+
+            if h.is_acceptable(role) {
+                h.handle(&req, &mut ctx, &mut acc).await?;
+            }
         }
         let header = std::mem::take(&mut acc.header);
         let res = HeartbeatResponse {
             header,
-            payload: acc.into_payload(),
+            mailbox_messages: acc.into_mailbox_messages(),
         };
         Ok(res)
     }
 }
+
+pub struct HeartbeatMailbox {
+    pushers: Arc<RwLock<BTreeMap<String, Pusher>>>,
+    sequence: Sequence,
+    senders: DashMap<MessageId, oneshot::Sender<Result<MailboxMessage>>>,
+    timeouts: DashMap<MessageId, Duration>,
+    timeout_notify: Notify,
+}
+
+impl HeartbeatMailbox {
+    pub fn create(
+        pushers: Arc<RwLock<BTreeMap<String, Pusher>>>,
+        sequence: Sequence,
+    ) -> MailboxRef {
+        let mailbox = Arc::new(Self::new(pushers, sequence));
+
+        let timeout_checker = mailbox.clone();
+        common_runtime::spawn_bg(async move {
+            timeout_checker.check_timeout_bg(1).await;
```

Comment:
Only 1 millisecond's interval?

Reply:
It's fine. 1 ms will make the timeout mechanism relatively more accurate. It won't consume excessive CPU usage, and if there is no mailbox RPC invoke, it will be blocked by the `notify`.

Reply:
But it still wastes CPU times on iterating the `timeouts` map if the deadlines are not reached. I think the timeout mechanism on this doesn't need to be that accurate, Let's make it 10ms at least.